subsys/bluetooth/audio/cap_initiator.c

```diff
@@ -378,22 +476,868 @@ int bt_cap_initiator_unicast_discover(struct bt_conn *conn)
+		return -EINVAL;
+	}
+
+	active_procedure.unicast_group = unicast_group;
+
+	return cap_initiator_unicast_audio_configure(param);
 }
 
-int bt_cap_initiator_unicast_audio_update(struct bt_audio_unicast_group *unicast_group,
-					  uint8_t meta_count,
-					  const struct bt_codec_data *meta)
+void bt_cap_initiator_codec_configured(struct bt_cap_stream *cap_stream)
 {
-	return -ENOSYS;
+	struct bt_conn *conns[MIN(CONFIG_BT_MAX_CONN,
+				  CONFIG_BT_AUDIO_UNICAST_CLIENT_GROUP_STREAM_COUNT)];
+	struct bt_audio_unicast_group *unicast_group;
+
+	if (!cap_stream_in_active_procedure(cap_stream)) {
+		/* State change happened outside of a procedure; ignore */
+		return;
+	}
+
+	if (active_procedure.subproc_type == CAP_UNICAST_SUBPROCEDURE_TYPE_RELEASE) {
+		/* When releasing a stream, it may go into the codec configured state if
+		 * the unicast server caches the configuration - We treat it as a release
+		 */
+		bt_cap_initiator_released(cap_stream);
+		return;
+	} else if (active_procedure.subproc_type != CAP_UNICAST_SUBPROCEDURE_TYPE_CODEC_CONFIG) {
+		/* Unexpected callback - Abort */
+		active_procedure.err = -EBADMSG;
+		atomic_set_bit(active_procedure.proc_state_flags,
+			       CAP_UNICAST_PROCEDURE_STATE_ABORTED);
+	} else {
+		active_procedure.stream_done++;
+
+		LOG_DBG("Stream %p configured (%zu/%zu streams done)",
+			cap_stream, active_procedure.stream_done,
+			active_procedure.stream_cnt);
+
+		if (active_procedure.stream_done < active_procedure.stream_cnt) {
+			/* Not yet finished, wait for all */
+			return;
+		}
+	}
+
+	if (atomic_test_bit(active_procedure.proc_state_flags,
+			    CAP_UNICAST_PROCEDURE_STATE_ABORTED)) {
+		if (active_procedure.stream_done == active_procedure.stream_initiated) {
+			cap_initiator_unicast_audio_start_complete();
+		}
+
+		return;
+	}
+
+	/* The QoS Configure procedure works on a set of connections and a
+	 * unicast group, so we generate a list of unique connection pointers
+	 * for the procedure
+	 */
+	(void)memset(conns, 0, sizeof(conns));
+	for (size_t i = 0U; i < active_procedure.stream_cnt; i++) {
+		struct bt_conn *stream_conn = active_procedure.streams[i]->bap_stream.conn;
+		struct bt_conn **free_conn = NULL;
+		bool already_added = false;
+
+		for (size_t j = 0U; j < ARRAY_SIZE(conns); j++) {
+			if (stream_conn == conns[j]) {
+				already_added = true;
+				break;
+			} else if (conns[j] == NULL && free_conn == NULL) {
+				free_conn = &conns[j];
+			}
+		}
+
+		if (already_added) {
+			continue;
+		}
+
+		__ASSERT(free_conn, "No free conns");
+		*free_conn = stream_conn;
+	}
+
+	/* All streams in the procedure share the same unicast group, so we just
+	 * use the reference from the first stream
+	 */
+	unicast_group = active_procedure.streams[0]->bap_stream.unicast_group;
+	active_procedure.stream_done = 0U;
+	active_procedure.stream_initiated = 0U;
+
+	active_procedure.subproc_type = CAP_UNICAST_SUBPROCEDURE_TYPE_QOS_CONFIG;
```

Comment:
```suggestion
static void cap_set_active_procedure(procedure)
{
	active_procedure.stream_done = 0U;
	active_procedure.stream_initiated = 0U;
	active_procedure.subproc_type = CAP_UNICAST_SUBPROCEDURE_TYPE_QOS_CONFIG;
}
```

`cap_set_active_procedure(CAP_UNICAST_SUBPROCEDURE_TYPE_QOS_CONFIG);`